In [125]:
#importing necessary packages for accessing the files from google drive
import os
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

In [126]:
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [127]:
#Accessing the files from drive using File ID
downloaded = drive.CreateFile({'id':'1ksi366OOgzLWB_P_0_fioCjGW8ml5HoP'}) # replace the id with id of file you want to access
downloaded.GetContentFile('titles.csv')

In [128]:
downloaded1 = drive.CreateFile({'id':'1b_aczX4naWHkCihcGSDPEUE1J1Sm7qtb'}) # replace the id with id of file you want to access
downloaded1.GetContentFile('cast.csv')

In [130]:
downloaded2 = drive.CreateFile({'id':'12S4kn6XHdKYo5cNhCIkR0W9qfwKT606P'}) # replace the id with id of file you want to access
downloaded2.GetContentFile('release_dates.csv')

In [ ]:
%matplotlib inline 
#Displays the output within the Notebook
#If only %matplotlib is used then the output is displayed in separate window
import pandas as pd

In [ ]:
#from IPython.core.display import HTML
#css = open('style-table.css').read() + open('style-notebook.css').read()
#HTML('<style>{}</style>'.format(css))

In [ ]:
titles = pd.read_csv('titles.csv')
titles.head()

title  year
0    Tasveer Mere Sanam  1996
1              Only You  1994
2  El pueblo del terror  1970
3               Machine  2007
4                  MARy  2008

In [ ]:
cast = pd.read_csv('cast.csv')
cast.head()

title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4  31.0
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

In [132]:
release = pd.read_csv('release_dates.csv')
release.head()

title  year      country        date
0   #73, Shaanthi Nivaasa  2007        India  2007-06-15
1                   #BKKY  2016     Cambodia  2017-10-12
2                 #Beings  2015      Romania  2015-01-29
3               #Captured  2017          USA  2017-09-05
4  #Ewankosau saranghaeyo  2015  Philippines  2015-01-21

### How many movies are listed in the titles dataframe?

In [141]:
titles['title'].count()

244914

There are 244914 movies in the titles dataframe

### What are the earliest two films listed in the titles dataframe?

In [ ]:
earliest = titles.sort_values('year', ascending = True)
earliest.head()

title  year
193222                                         Miss Jerry  1894
51388                                  The Startled Lover  1898
4584       Reproduction of the Corbett and Jeffries Fight  1899
219198                                     May Day Parade  1900
47548   Pierrot's Problem, or How to Make a Fat Wife O...  1900

The earliest two movies listed in titles dataframe are:
1.   Miss Jerry
2.   The Startled Lover



### How many movies have the title "Hamlet"?

In [134]:
hamlet=titles[titles['title'].str.contains('Hamlet')]
hamlet.head()

title  year
6422               Hamlet  2017
10637              Hamlet  1921
39916              Hamlet  1913
46189       Zombie Hamlet  2012
48703  Han, hun og Hamlet  1922

In [136]:
hamlet['title'].count()

57

There are 57 movies having the title "Hamlet"

### How many movies are titled "North by Northwest"?

In [140]:
titles[titles['title'] == 'North by Northwest']['title'].count()

1

There is only one movie titled "North by Northwest"

### When was the first movie titled "Hamlet" made?

In [ ]:
hamlet[hamlet.title == 'Hamlet'].sort_values('year', ascending = True).head()

title  year
52521   Hamlet  1910
198234  Hamlet  1911
39916   Hamlet  1913
10637   Hamlet  1921
101728  Hamlet  1948

The first movie titled "Hamlet" was made in 1910

### List all of the "Treasure Island" movies from earliest to most recent.

In [ ]:
titles[titles.title == 'Treasure Island'].sort_values('year', ascending = True) 

title  year
55686   Treasure Island  1918
196514  Treasure Island  1920
141822  Treasure Island  1934
77354   Treasure Island  1950
65051   Treasure Island  1972
37238   Treasure Island  1973
173051  Treasure Island  1985
27076   Treasure Island  1999

*   The earliest movie of the 'Treasure Island' was released in the year 1918.
*   The latest movie of the 'Treasure Island' was released in the year 1999.




### How many movies were made in the year 1950?

In [ ]:
titles[titles.year == 1950].head()

title  year
189  La soif des hommes  1950
268    Sohrab at Rustum  1950
308               Gemma  1950
314        Rasuna valea  1950
524  Trial Without Jury  1950

In [122]:
titles[titles.year == 1950]['year'].count()

1113

There were 1113 movies made in year 1950

### How many movies were made in the year 1960?

In [ ]:
titles[titles.year == 1960].head()

title  year
52     Tayna Dimki Karmiya  1960
88   Tres angelitos negros  1960
226           Viuda de oro  1960
227        Tao se feng yun  1960
590   Mahallenin sevgilisi  1960

In [123]:
titles[titles.year == 1960]['year'].count()

1549

There were 1549 movies made in the year 1960

### How many movies were made from 1950 through 1959?

In [ ]:
titles[(titles['year'] >= 1950) & (titles['year'] < 1960)]

title  year
5                    Tight Spot  1955
8               Bao dao mei ren  1956
45           Trois de la marine  1957
48                    Hong hong  1956
61              Alarma in munti  1955
...                         ...   ...
244803             Touch and Go  1955
244817            White Feather  1955
244818      Carovana di canzoni  1955
244820                 Doctor X  1950
244841  Matros soshyol na bereg  1957

[12934 rows x 2 columns]

In [ ]:
len(titles[(titles['year'] >= 1950) & (titles['year'] < 1960)])

12934

The total number of movies released between 1950 to 1960 is 12934

### In what years has a movie titled "Batman" been released?

In [ ]:
titles[titles.title == 'Batman']

title  year
39349   Batman  1943
240896  Batman  1989

In years 1943 and 1989 the Batman movie has been released

In [ ]:
cast.head()

title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4  31.0
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

### How many roles were there in the movie "Inception"?

In [121]:
inception = cast[cast.title == 'Inception']
inception['character'].count()

78

There are total 78 roles in the movie "Inception"

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [ ]:
inception['n'].isna().sum()

27

There are 27 'n' values which are not ranked in the movie 'Inception'

### But how many roles in the movie "Inception" did receive an "n" value?

In [119]:
 inception[pd.notnull(inception['n'])]['n'].count()

51

There are 51 roles in the movie "Inception" which received 'n' values

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [120]:
#Sorting the values according to the 'n'-value order
north = cast[cast.title == 'North by Northwest'].sort_values('n', ascending = True)

#ignoring the values which do not have a 'n' value
#Hence filtering the whole dataframe
bool_series = pd.notnull(north['n'])

#Displaying the cast of "North by Northwest"
north[bool_series]['name']

871426               Cary Grant
3518333         Eva Marie Saint
1457165             James Mason
3157232     Jessie Royce Landis
356707           Leo G. Carroll
3048272    Josephine Hutchinson
1695173             Philip Ober
1275337           Martin Landau
2447081           Adam Williams
1811387            Edward Platt
665768        Robert Ellenstein
2295714            Les Tremayne
464736          Philip Coolidge
1508344           Patrick McVey
206471             Edward Binns
1384780               Ken Lynch
Name: name, dtype: object

### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [100]:
cast[(cast.title == "Sleuth") & (cast.year == 1972)].sort_values('n', ascending = True)
#year_1972 = sleuth[sleuth.year == 1972]
#year_1972

title  year  ...                   character    n
1705394  Sleuth  1972  ...                 Andrew Wyke  1.0
326773   Sleuth  1972  ...                 Milo Tindle  2.0
373876   Sleuth  1972  ...           Inspector Doppler  3.0
1465483  Sleuth  1972  ...  Detective Sergeant Tarrant  4.0
2724553  Sleuth  1972  ...             Marguerite Wyke  5.0
1448690  Sleuth  1972  ...      Police Constable Higgs  6.0

[6 rows x 6 columns]

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [ ]:
year_2007 = sleuth[sleuth.year == 2007]
year_2007

title  year                 name     type          character    n
326774   Sleuth  2007        Michael Caine    actor             Andrew  1.0
1293634  Sleuth  2007             Jude Law    actor               Milo  2.0
1805907  Sleuth  2007        Harold Pinter    actor        Man on T.V.  3.0
260242   Sleuth  2007      Kenneth Branagh    actor  Other Man on T.V.  NaN
373877   Sleuth  2007  Alec (II) Cawthorne    actor  Inspector Doppler  NaN
2724552  Sleuth  2007    Eve (II) Channing  actress    Marguerite Wyke  NaN
3370424  Sleuth  2007    Carmel O'Sullivan  actress             Maggie  NaN

### How many roles were credited in the silent 1921 version of Hamlet?

In [ ]:
#silent = cast[cast.title == "Hamlet"].sort_values('n', ascending = True)
#silent[silent.year == 1921]
#silent[silent.year == 1921].count(axis = 0)
len(cast[(cast.title == 'Hamlet') & (cast.year == 1921)].character)

9

A total of 9 roles were credited in the silent 1921 version of Hamlet

### How many roles were credited in Branagh’s 1996 Hamlet?

In [ ]:
#role = cast[cast.title == "Hamlet"].sort_values('n', ascending = True)
#role[role.year == 1996].count(axis=0)
len(cast[(cast.title == 'Hamlet') & (cast.year == 1996)].character)

54

There are total 54 roles which were credited in Branagh's 1996 Hamlet

### How many "Hamlet" roles have been listed in all film credits through history?

In [ ]:
#roles = cast[cast.title == "Hamlet"].sort_values('n', ascending = True)
#roles.count()
len(cast[cast.title == 'Hamlet'])

342

There are total 342 roles in "Hamlet" which have been listed in all film credits through history

### How many people have played an "Ophelia"?

In [ ]:
len(cast[cast['character'] == 'Ophelia'].character)

117

There are total 117 people who have played as "Ophelia"

### How many people have played a role called "The Dude"?

In [ ]:
len(cast[cast['character'] == 'The Dude'].character)

19

There are 19 people who have played the tole of "The Dude"

### How many people have played a role called "The Stranger"?

In [ ]:
len(cast[cast['character'] == 'The Stranger'].character)

212

There were 212 people who have played a role called 'The Stranger'

### How many roles has Sidney Poitier played throughout his career?

In [ ]:
len(cast[cast['name'] == 'Sidney Poitier'].name)

43

Sidney Poiter played 43 roles throughout his career

### How many roles has Judi Dench played?

In [ ]:
len(cast[cast['name'] == 'Judi Dench'].name)

55

Judi Dench has played 55 roles 

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [ ]:
cary = cast[(cast.name == 'Cary Grant') & (cast.year >= 1940)].sort_values('year', ascending = True)
cary = cary[cary['n'].notnull()]
#cary = cary[cary.n == 2].sort_values('year', ascending = True)
cary[(cary['n'] == 2) & (cary['year'] <= 1950)]['character']

871423     Nick Arden
871433    Roger Adams
Name: character, dtype: object

There are only 2 supporting roles perfromed by Cary Grant in the 1940s

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [ ]:
lead = cast[(cast.name == 'Cary Grant') & (cast.year >= 1940)].sort_values('year', ascending = True)
lead = lead[lead['n'].notnull()]
lead[(lead['n'] == 1) & (lead['year'] <= 1950)]

title  ...    n
871407                       His Girl Friday  ...  1.0
871450                The Philadelphia Story  ...  1.0
871448               The Howards of Virginia  ...  1.0
871438                             Suspicion  ...  1.0
871429                 Once Upon a Honeymoon  ...  1.0
871452                  The Talk of the Town  ...  1.0
871422                             Mr. Lucky  ...  1.0
871398                     Destination Tokyo  ...  1.0
871390                  Arsenic and Old Lace  ...  1.0
871430                      Once Upon a Time  ...  1.0
871425             None But the Lonely Heart  ...  1.0
871424                         Night and Day  ...  1.0
871427                             Notorious  ...  1.0
871444      The Bachelor and the Bobby-Soxer  ...  1.0
871445                     The Bishop's Wife  ...  1.0
871402          Every Girl Should Be Married  ...  1.0
871421  Mr. Blandings Builds His Dream House  ...  1.0
871411                I Was a Male War Bride  ...  1.0
871396                                Crisis  ...  1.0

[19 rows x 6 columns]

There are 19 leading roles played by Cary Grant in 1940s

### How many roles were available for actors in the 1950s?

In [99]:
actor = cast[(cast.year >= 1950) & (cast.year <= 1960)]
actor[actor['type'] == 'actor']['year'].count()

171302

There were 171302 roles available for actors in the 1950s

### How many roles were available for actresses in the 1950s?

In [97]:
actress = cast[(cast.year >= 1950) & (cast.year <= 1960)]
actress[actress['type'] == 'actress']['year'].count()

63333

There were 63333 roles available for actresses in the 1950s

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [ ]:
len(cast[(cast.n == 1) & (cast.year >= 1980)])

89276

There were as many as 89726 leading roles available from the beginning of film history through 1980

### How many non-leading roles were available through from the beginning of film history through 1980?

In [96]:
#non_lead = cast[cast['n'].notnull()].sort_values('year')
len(non_lead[(non_lead.n != 1) & (non_lead.year >= 1980)])

1522873

The total number of non-leading roles that were available were 1522873 from the beginning of film history through 1980

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [95]:
len(cast[(cast['n'].isnull()) & (cast['year'] >= 1980)])

1014692

There were as many as 1014692 roles which did not warant a numeric 'n' rank through 1980